### DISCUSSION TITLE
RPC Order of Operation

### DISCUSSION THREAD

**Discussion by lewisxy** [capnproto/capnproto#1862] [Q&A] [answered]
Suppose I made several RPC calls, are they always received in the order they are sent? Are they always processed in the order they are received?

For example, consider the following client code.
```c++
MyInterface::Client apiClient = ...; // bootstrap the main interface using ez-rpc or TwoPartyClient
MyPromiseContainer container; // some container of promises
for (int i = 0; i < 100; ++i) {
  {
    auto req = apiClient.myMethodRequest();
    req.setI(i);
    container.add(req.send()); // no wait
  }
  {
    auto req = apiClient.myMethod2Request();
    req.setI(i);
    container.add(req.send()); // no wait
  }
}
container.waitAll(...); // wait for everything inside
```
Is it guaranteed that server received call in the following order?
```
myMethod(0), myMethod2(0), myMethod(1), myMethod2(1), ..., myMethod(99), myMethod2(99)
```

**Comment by kentonv** [comment, marked as answer]
If all the calls are made on the same `Client` object, then they are guaranteed to be delivered in the order they are sent. This property is known as "e-order" (from the E programming language, which introduced it). The RPC system goes to great lengths to preserve e-order as many applications depend on it.

Note that there is no guarantee of ordering when calls are made on different `Client`s, even if they point to the same destination object. (Although, in practice, calls made from the same client thread to the same server thread will usually preserve ordering even if the calls are on different objects.)
